homeassistant/components/sensor/netdata.py

```diff
@@ -4,154 +4,152 @@
-    'system_io_out': ['System IO Out', 'Count', 'system.io', 'out', 0],
-    'ipv4_in': ['IPv4 In', 'kb/s', 'system.ipv4', 'received', 0],
-    'ipv4_out': ['IPv4 Out', 'kb/s', 'system.ipv4', 'sent', 0],
-    'disk_free': ['Disk Free', 'GiB', 'disk_space._', 'avail', 2],
-    'cpu_iowait': ['CPU IOWait', '%', 'system.cpu', 'iowait', 1],
-    'cpu_user': ['CPU User', '%', 'system.cpu', 'user', 1],
-    'cpu_system': ['CPU System', '%', 'system.cpu', 'system', 1],
-    'cpu_softirq': ['CPU SoftIRQ', '%', 'system.cpu', 'softirq', 1],
-    'cpu_guest': ['CPU Guest', '%', 'system.cpu', 'guest', 1],
-    'uptime': ['Uptime', 's', 'system.uptime', 'uptime', 0],
-    'packets_received': ['Packets Received', 'packets/s', 'ipv4.packets',
-                         'received', 0],
-    'packets_sent': ['Packets Sent', 'packets/s', 'ipv4.packets',
-                     'sent', 0],
-    'connections': ['Active Connections', 'Count',
-                    'netfilter.conntrack_sockets', 'connections', 0]
-}
+DEFAULT_PORT = 19999
+
+DEFAULT_ICON = 'mdi:desktop-classic'
+
+RESOURCE_SCHEMA = vol.Any({
+    vol.Required(CONF_ELEMENT): cv.string,
+    vol.Optional(CONF_ICON, default=DEFAULT_ICON): cv.icon,
+    vol.Optional(CONF_NAME): cv.string,
+}, None)
 
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
     vol.Optional(CONF_HOST, default=DEFAULT_HOST): cv.string,
     vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
     vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
-    vol.Optional(CONF_RESOURCES, default=['memory_free']):
-        vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+    vol.Required(CONF_RESOURCES): vol.Schema({cv.string: RESOURCE_SCHEMA}),
 })
 
 
-# pylint: disable=unused-variable
-def setup_platform(hass, config, add_devices, discovery_info=None):
+async def async_setup_platform(
+        hass, config, async_add_devices, discovery_info=None):
     """Set up the Netdata sensor."""
+    from netdata import Netdata
+
     name = config.get(CONF_NAME)
     host = config.get(CONF_HOST)
     port = config.get(CONF_PORT)
-    url = 'http://{}:{}'.format(host, port)
-    data_url = '{}/{}/data?chart='.format(url, _RESOURCE)
     resources = config.get(CONF_RESOURCES)
 
-    values = {}
-    for key, value in sorted(SENSOR_TYPES.items()):
-        if key in resources:
-            values.setdefault(value[2], []).append(key)
+    session = async_get_clientsession(hass)
+    netdata = NetdataData(Netdata(host, hass.loop, session, port=port))
+    await netdata.async_update()
+
+    if netdata.api.metrics is None:
+        raise PlatformNotReady
 
     dev = []
-    for chart in values:
-        rest_url = '{}{}&{}'.format(data_url, chart, _REALTIME)
-        rest = NetdataData(rest_url)
-        rest.update()
-        for sensor_type in values[chart]:
-            dev.append(NetdataSensor(rest, name, sensor_type))
+    for entry, data in resources.items():
+        sensor = entry
+        element = data.get(CONF_ELEMENT)
```

Comment:
This will error if `data` is `None`.

Reply:
Ups, this should never be `None` as `element:` should be required.

Reply:
If element is required, use `data[CONF_ELEMENT]` to indicate in the code that we assume it's there.

Reply:
We don't need to assume that it's there, the config validation make sure that it's present.